### ISSUE TITLE
Add support for custom usage of unary operators

### ISSUE THREAD

**Issue by TDA** [jscs-dev/node-jscs#1511]
Would it be possible to add support for custom usage of unary operators? especially the negation operator.
Example:
We use stuff like "!!(random condition)" in a lot of places on the @Mozilla fxa-content-server repos. We also use "!condition" at some places. So, we want JSCS to throw an error when there is no space after the "!" unary operator, but NOT for usages like the "!!".
Preferred approach would be something like,
"disallowSpaceAfterPrefixUnaryOperators": ["++", "--", "+", "-". "!!"],
"requireSpaceAfterPrefixUnaryOperators": ["~", "!"]

However, this is currently not supported by JSCS. Would it be possible to add this?
https://github.com/mozilla/fxa-content-server/pull/2695
@pdehaan

**Comment by hzoo** [comment]
Not really sure what your asking about - looks like there is http://jscs.info/rule/requireSpaceAfterPrefixUnaryOperators.html and http://jscs.info/rule/disallowSpaceAfterPrefixUnaryOperators.html already

**Comment by TDA** [comment]
Uhm, I am asking about custom usage of unary operators, like "!!", pretty positive that isn't in the docs. I have described the use case in detail above, with an example as well.

**Comment by pdehaan** [comment]
@hzoo Here is a simple use case:

#### index.js
```js
'use strict'

if (!false) {
  console.log(true)
}

if (! true) {
  console.log(false)
}

if (!!'foo') {
  console.log('foo')
}

if (!! 'bar') {
  console.log('bar')
}
```

#### .jscsrc
```json
{
  "disallowSpaceAfterPrefixUnaryOperators": ["!!"],
  "requireSpaceAfterPrefixUnaryOperators": ["!"]
}
```

#### Output:
```sh
$ jscs --version # 1.9.0
$ jscs .
Operator ! should not stick to operand at ./index.js :
     1 |'use strict'
     2 |
     3 |if (!false) {
------------^
     4 |  console.log(true)
     5 |}

Operator ! should not stick to operand at ./index.js :
     9 |}
    10 |
    11 |if (!!'foo') {
-------------^
    12 |  console.log('foo')
    13 |}


2 code style errors found.
```

I think the first error is valid. I specified `"requireSpaceAfterPrefixUnaryOperators": ["!"]` so there should be an error and I should have typed `if (! false) {...}`.

I think the second error is incorrect. I specified `"disallowSpaceAfterPrefixUnaryOperators": ["!!"],` so that should be valid syntax, and I would have expected an error on the other example since it _does_ have a space after the `!!`: `if (!! 'bar') { ... }`.

**Comment by pdehaan** [comment]
Same results with `v1.13.1`...

I think this pretty much sums it up though:
```
Operator ! should not stick to operand at ./index.js :
    11 |if (!!'foo') {
-------------^
    12 |  console.log('foo')
    13 |}
```

Note that the error message says "Operator ! should not stick to operand at ./index.js : `if (!!'foo') {...}`"
It says "Operator `!` but the actual operator is `!!`. I'm guessing that the `requireSpaceAfterPrefixUnaryOperators` rule needs an additional check to make sure that the `!` operator it isn't actually an `!!` operator.

**Comment by markelog** [comment]
Sounds like a different rule, we will accept it, if you choose to send us a pr

**Comment by pdehaan** [comment]
@markelog I'm pretty sure it's a bug in the `requireSpaceAfterPrefixUnaryOperators` rule.
If we specify that we want a space after a `!` but not `!!`, it still parses `!!` and tries to enforce a space.

**Comment by markelog** [comment]
Well, there is no operator `!!` there is only `!` this is why it parsed like that

**Comment by hzoo** [comment]
Ok the rule has code in it related to this - https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/require-space-after-prefix-unary-operators.js#L66 and related issue https://github.com/jscs-dev/node-jscs/issues/405

**Comment by TDA** [comment]
@markelog: You are right about that, but this is a very 'nice-to-have' functionality that cannot be implemented without adding some logic to the token parsing. So would it be possible for you guys to help us out with this? I see a couple of issues very similar to this one that have been resolved in the past. We would appreciate it if you could help us on this one.

**Comment by markelog** [comment]
I will put that on my list, but it will be much faster if you choose to send us a PR for the new rule (preferable for the opposite one too). 

fyi we're planing to release 2.0 on monday.

**Comment by osv** [comment]
+1 this may be need if you use [Meteor-Style-Guide#spaces-between-tokens](https://github.com/meteor/meteor/wiki/Meteor-Style-Guide#spaces-between-tokens)

**Comment by hzoo** [comment]
We released [3.0](https://github.com/jscs-dev/node-jscs/releases/tag/v3.0.0) recently! We also wrote a [post](https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2#.oot3wvgpu) about how 3.0 will be our last major release (since we are deprecating the project) and that we are [merging with ESLint](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint)!

Thus we won't be working on any new features moving forward which includes new rules/options. We will continue to fix bugs and work on CST/better autofixing for current rules in the upcoming months both to help JSCS users and to gain experience that we can hopefully use in ESLint. But otherwise we both teams will be working on getting [ESLint feature parity with JSCS](https://github.com/eslint/eslint/milestones/JSCS%20Compatibility) and just a nice transition for our users.

I'm putting an `orphaned` label to note this issue isn't planned to be worked on, but might be a good contribution as a custom rule (not in JSCS core itself).

I would encourage anyone who wants the rule to try writing it themselves: I would either write a custom rule or submit an issue to ESLint and if accepted a PR. Otherwise a custom rule that won't be in core repo itself.

JSCS rules are also very similar to ESLint rules: here are some resources to get started: 

- [Project-specific Lint Rules with ESLint](https://gist.github.com/jareware/7179093)
- [Creating Custom Rules for ESLint](https://insideops.wordpress.com/2015/12/08/creating-custom-rules-for-eslint)
- [Maintaining JavaScript Code Quality with ESLint ](https://www.paypal-engineering.com/2014/12/12/maintaining-javascript-code-quality-with-eslint/)
